From 1df05ac832e392c2b08286f11b3e01ad9ef1d3d2 Mon Sep 17 00:00:00 2001
From: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Date: Fri, 7 Oct 2022 13:35:59 +1100
Subject: normal/menu: Add Ctrl-L to refresh the menu

This is useful on cloud instances with remote serial ports as it can be
difficult to connect "fast enough" to get the initial menu display

Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>
---
 docs/grub.texi          | 3 +++
 grub-core/normal/menu.c | 4 ++++
 2 files changed, 7 insertions(+)

diff --git a/docs/grub.texi b/docs/grub.texi
index 211cddc..0dbbdc3 100644
--- a/docs/grub.texi
+++ b/docs/grub.texi
@@ -3222,6 +3222,9 @@ If you protect the menu interface with a password (@pxref{Security}),
 all you can do is choose an entry by pressing @key{RET}, or press
 @key{p} to enter the password.
 
+Pressing @key{Ctrl-l} will refresh the menu, which can be useful when
+connecting via serial after the menu has been drawn.
+
 
 @node Menu entry editor
 @section Editing a menu entry
diff --git a/grub-core/normal/menu.c b/grub-core/normal/menu.c
index fa4871a..6a90e09 100644
--- a/grub-core/normal/menu.c
+++ b/grub-core/normal/menu.c
@@ -788,6 +788,10 @@ run_menu (grub_menu_t menu, int nested, int *auto_boot, int *notify_boot)
 		}
 	      goto refresh;
 
+	    case GRUB_TERM_CTRL | 'l':
+	      menu_fini ();
+	      goto refresh;
+
 	    default:
 	      {
 		int entry;
-- 
cgit v1.1

